src/theory/strings/inference_manager.cpp

```diff
@@ -320,23 +274,25 @@ void InferenceManager::doPendingFacts()
   size_t i = 0;
   while (!d_state.isInConflict() && i < d_pending.size())
   {
-    Node fact = d_pending[i].d_fact;
-    Node exp = d_pending[i].d_exp;
-    if (fact.getKind() == AND)
-    {
-      for (const Node& lit : fact)
-      {
-        bool polarity = lit.getKind() != NOT;
-        TNode atom = polarity ? lit : lit[0];
-        assertPendingFact(atom, polarity, exp);
-      }
-    }
-    else
-    {
-      bool polarity = fact.getKind() != NOT;
-      TNode atom = polarity ? fact : fact[0];
-      assertPendingFact(atom, polarity, exp);
-    }
+    InferInfo& ii = d_pending[i];
+    // At this point, ii should be a "fact", i.e. something whose conclusion
+    // should be added as a normal equality or predicate to the equality engine
+    // with no new external assumptions (ii.d_antn).
+    Assert(ii.isFact());
+    Node fact = ii.d_conc;
+    Node exp = utils::mkAnd(ii.d_ant);
+    Trace("strings-assert") << "(assert (=> " << exp << " " << fact
+                            << ")) ; fact " << ii.d_id << std::endl;
+    // only keep stats if we process it here
+    Trace("strings-lemma") << "Strings::Fact: " << fact << " from " << exp
+                           << " by " << ii.d_id << std::endl;
+    d_statistics.d_inferences << ii.d_id;
+    // assert it as a pending fact
+    bool polarity = fact.getKind() != NOT;
+    TNode atom = polarity ? fact : fact[0];
+    // no double negation or double (conjunctive) conclusions
+    Assert(atom.getKind() != NOT && atom.getKind() != AND);
```

Comment:
Should we maybe add these conditions to `InferInfo::isFact()`? I find it a bit confusing that something can be a fact but then can't be sent out as a fact.

Reply:
There is a distinction here about whose responsibility it is to use the class correctly.  Since this data structure is check is done millions of times, it would be nice to have some basic contracts on what is allowed to be a valid infer info.  The first check here is definitely a sanity check: no one should make a conclusion of the form (not (not F)).  The second is debatable, an earlier version of this PR had this check for AND in isFact.  I'll think about it a bit more.

Reply:
We've discussed the handling of facts/lemmas offline and will potentially revisit it in a future PR.